homeassistant/components/rointe/device_manager.py

```diff
@@ -0,0 +1,384 @@
+            "Product not found: [%s][%s] at v[%s]",
+            product_type,
+            version,
+            current_firmware,
+        )
+        return None
+
+    if product in fw_map and version in fw_map[product]:
+        return fw_map[product][version]
+
+    # If no update path available return the current firmware string.
+    return current_firmware
+
+
+class RointeDeviceManager:
+    """Device Manager."""
+
+    def __init__(
+        self,
+        username: str,
+        password: str,
+        installation_id: str,
+        hass: HomeAssistant,
+        rointe_api: RointeAPI,
+    ) -> None:
+        """Initialize the device manager."""
+        self.username = username
+        self.password = password
+        self.installation_id = installation_id
+        self.rointe_api = rointe_api
+
+        self.hass = hass
+        self.auth_token = None
+        self.auth_token_expire_date: datetime | None = None
+
+        self.rointe_devices: dict[str, RointeDevice] = {}
+
+    def _fail_all_devices(self):
+        """Set all devices as unavailable."""
+
+        if self.rointe_devices:
+            for device in self.rointe_devices.values():
+                device.hass_available = False
+
+    async def update(self) -> dict[str, list[RointeDevice]]:
+        """Retrieve the devices from the user's installation."""
+
+        LOGGER.debug("Device manager updating")
+
+        installation_response: ApiResponse = await self.hass.async_add_executor_job(
+            self.rointe_api.get_installation_by_id, self.installation_id
+        )
+
+        if not installation_response.success:
+            LOGGER.error(
+                "Unable to get Rointe Installations. Error: %s",
+                installation_response.error_message,
+            )
+            self._fail_all_devices()
+            return {}
+
+        installation = installation_response.data
+        discovered_devices: dict[str, list[RointeDevice]] = {}
+        firmware_map = await self._get_firmware_map()
+
+        for zone_key in installation["zones"]:
+            LOGGER.debug("Processing zone: %s", zone_key)
+
+            zone = installation["zones"][zone_key]
+
+            if "devices" not in zone:
+                LOGGER.debug("No devices info found for zone")
+                continue
+
+            devices = zone["devices"]
+
+            for device_id in devices:
+                LOGGER.debug("Processing device ID: %s", device_id)
+
+                if not devices[device_id]:
+                    LOGGER.debug("Device ID: %s has no data", device_id)
+
+                    continue
+
+                device_data_response: ApiResponse = (
+                    await self.hass.async_add_executor_job(
+                        self.rointe_api.get_device, device_id
+                    )
+                )
+
+                if device_data_response.success:
+                    # Retrieve energy stats.
+                    energy_stats_response: ApiResponse = (
```

Comment:
We're looping through devices and making a lot of synchronous executor jobs—can they be collected and batched?

You might create a separate method that gets device data and, if successful, gets energy stats data—then, batch _that_ method together.

Reply:
I don't quite understand what you mean by batching the methods.

Are you referring to iterating in parallel (akin to `joblib.Parallel` )?

Or are you referring to making a single call to get all the device's data and then another call to get all device's energy data? Unfortunately the API doesn't allow me to query all devices in a single call. It's possible with energy data but I have to request a node so high up on their object hierarchy that the response would be massive and would take > 300s to return.

Or something else? :)

Reply:
I'm referring to the former more than the latter.

`hass.async_add_executor_job` creates an asynchronous task (really, an asynchronous wrapper and a thread inside which synchronous code is run). You can run multiple asynchronous tasks simultaneously via the `asyncio.gather` call.

As a (overly simple) example, imagine this:

```python
def test() -> None:
    """Run."""
    print("Running the test function")

tasks = []
for i in range(0, 5):
    tasks.append(self.hass.async_add_executor_job(test))
 
await asyncio.gather(*tasks)
```

This will run five iterations of the `test` function in parallel rather than one-by-one. This silly example won't make much difference, but in your world (where API calls are performing I/O), it can make a big difference.

Of course, this assumes that each task _can_ be run in parallel. If you have API calls that _have_ to happen before others, you have to consider whether this can be revised. For example, if we assume that a different function—`test2`—needs to run before `test`, but still want to explore batching things, we might consider:

```python
def test() -> None:
    """Run."""
    print("Running the test function")

def test2() -> None:
    """Run."""
    print("Running the test2 function")

def combined() -> None:
    """Run."""
    test()
    test2()

tasks = []
for i in range(0, 5):
    tasks.append(self.hass.async_add_executor_job(combined))
 
await asyncio.gather(*tasks)
```

Reply:
Got it. I think I can make this work!

Reply:
Refactored the code and split the logic through a couple of methods. Please tell me if you approve this approach.

Also, do I need to control the degree of parallelism? Does HASS manages that for me?